Load previously lost decorator code and don't delete orders

mkllnk opened the pull request:
#### What? Why?

While investigating #5897 I found some code that could accidentally delete an order. But then I found that the code wasn't loaded and could not have caused a disappearing order.

I continued to investigate the issue of unloaded code because it could be responsible of many random issues. The deeper I dug, the clearer I could see that the unloaded code was mostly insignificant. Nonetheless, I addressed the loading issue and removed the code that could delete orders.

The non-loading was mostly no issue because our code didn't change Spree's logic. The only exception are the changes from 20f610fbee4f6a9e22bb77e3baa7fc6cb684a628 which were lost since v3 and now brought back.

#### What should we test?
<!-- List which features should be tested and how. -->

##### Parallel browsing

* As guest user, go to a shop and add items to the cart.
* Open a different browser (or a private window or a different device) and do the same there with different items in your cart.
* The two carts should be persisted and independent.
* Log in with the same user in both browser sessions.
* Both sessions should keep their individual orders.
* Go through checkout with the first browser session.
* The first browser should now show an empty cart again.
* The second browser should still have its old cart.



#### Release notes
<!-- Write a line or two to be included in the release notes.
Everything is worth mentioning, because you did it for a reason. -->



<!-- Please assign one category to your PR and delete the others. 
The categories are based on https://keepachangelog.com/en/1.0.0/. -->

Changelog Category: Added | Changed | Deprecated | Removed | Fixed | Security

#### How is this related to the Spree upgrade?
<!-- Any known conflicts with the Spree Upgrade?
Explain them or remove this section. -->



#### Discourse thread
<!-- Is there a discussion about this in Discourse?
Add the link or remove this section. -->



#### Dependencies
<!-- Does this PR depend on another one?
Add the link or remove this section. -->



#### Documentation updates
<!-- Are their any wiki pages that need updating after merging this PR?
List them here or remove this section. -->

luisramos0 commented on lib/open_food_network/model_class_from_controller_name.rb:
```diff
@@ -1,8 +0,0 @@
-      self.class.to_s.sub("Controller", "").underscore.split('/').last.singularize.to_sym
-    end
-  end
-end
```
:heart:

luisramos0 commented on spec/lib/spree/core/environment_spec.rb:
```diff
@@ -0,0 +1,19 @@
+
+describe Spree::Core::Environment do
+  # Our version doesn't add any features we could test.
+  # So we just check that our file is loaded correctly.
```
:+1:
Nice one.
I think I moved the file to remove the stock_splitters attribute.

luisramos0 commented on app/controllers/base_controller.rb:
```diff
@@ -1,3 +1,7 @@
+require 'spree/core/controller_helpers'
```
I dont think we need this line and line 9 below.
the 4 helpers in ControllerHelpers are included individually.

luisramos0 commented:
Note: on all investigations I have done related to failed payments I only found missing entries on spree_payments. spree_orders were always there. So:
- option 1 - the payments issues I investigated are different from  #5897 where the order is missing
- option 2 - the order.destroy that we are removing here could, in some cases, destroy the payments but not the order?
I'd say option 1 is more likely and this is a different problem from what we have done in #5080 for example.

luisramos0 commented:
I'd mark this as closing #5897 :muscle: 
I'd merge this, close #5897 and then reopen 5897 if we do have any other occurrences.

luisramos0 commented:
I think this could also have an impact on #5862 
I think we need to get this PR in the next release, ideally merge before this Thursday.

mkllnk commented:
Thanks for your early review. I'll continue my work today. I should have mentioned on Friday that I left it as draft because I didn't verify all my claims. I didn't go through the testing steps myself, they are just guesses from what I saw in the code. At the weekend, I also remembered that some controllers did load the helper file (CartController and Spree::OrdersController). So the deletion of old carts could have happened when logging in on the cart page, maybe? I'll try to replicate.

luisramos0 commented:
ok :+1:
Will you manage to finish this tomorrow?
I think this PR, as is, can fix the bug and is harmless so I think we should go ahead with it in this weeks release :+1:

luisramos0 commented:
Hey @mkllnk I think now I remember in delivery train meeting you suggesting I pick this one up so you can work on mobile?
I am correct? 
If so, what exactly do you think still needs to be done here? Can we just move this to Test Ready?

mkllnk commented:
I needed to verify the test instructions and found that one part of it was invalid. I added another two commits, addressing your feedback, @luisramos0, and removing the dead code. There is more I can clean up but I will do that in a separate PR.

mkllnk commented:
@luisramos0 I added two more commits in a separate PR: #5927 If you think it's better to include them here and we can still make it for the release, that would be awesome but I didn't want to put anything in the way of this PR getting into the release.

luisramos0 commented on lib/spree/core/controller_helpers/order.rb:
```diff
@@ -68,18 +68,9 @@ def associate_user
-
-          last_incomplete_order = user.last_incomplete_spree_order
-
-          if session[:order_id].nil? && last_incomplete_order
```
I am not sure about this statement:
"Every page load creates a cart order if none is present."

I thought this method was the place where that was happening, right?
I thought that this conditional was recovering existing carts if no order is identified in the session.
By doing this I think we are just increasing the number of past carts for the users and even maybe loosing resetting the cart between sessions.

My test would be: login to OFN and add something to the cart.
Then, open an icognito tab and login again, do you see your cart as it was?

luisramos0 replied:
Yeah, my hypothesis is confirmed by my test.
I run this test above in master and with this PR and in master the cart is preserved into the incognito tab after login, with this PR, after login in the incognito tab, the cart is empty.
We need to revert this change and keep this conditional.

luisramos0 replied:
I reverted the last commit so we can move this PR forward now.

Matt-Yorkley replied:
Sounds good to me

luisramos0 replied:
ok, awesome, moving to Test Ready now.

mkllnk replied:
Okay, great. I tried to test that and I couldn't find a way to recover an old cart. Can you share your steps?

luisramos0 replied:
just these as described above:
- login to OFN and add something to the cart.
- Then, open an icognito tab and login again, do you see your cart as it was?

filipefurtad0 commented:
Hi @mkllnk ,

Thanks for this investigation and the detailed testing-steps!

Since master branch appears to be broken (as discussed here), I have staged the build used for release testing (v3.2.3), before testing this PR. The images below show session 1 on (the left) and session 2 (on the right side). I followed your procedure:


i) As guest user, went to a shop and added 5 items to the cart.
ii) Repeat the step above, on a different browser/private window. (I found no to play a role, if they are the same items or not)

At this point, both guest sessions have 5 items each. Proceeded:

iii) Logged in with the same user in both sessions

Here it became clear, that the condition
"- Both sessions should keep their individual orders." - is not verified.

The second session, now contains 10 items, the items from session 1 plus its own:

![image](https://user-images.githubusercontent.com/49817236/90732504-19f1d280-e2c3-11ea-88dd-fddbad624086.png)


iv) At this point, attempting to checkout on session 1 generates the error:

![image](https://user-images.githubusercontent.com/49817236/90732566-32fa8380-e2c3-11ea-9ee0-a550c416c69b.png)

So,
- "The first browser should now show an empty cart again." - is not verified
-  "The second browser should still have its old cart." - it still does, but it contains the items from session 1 as well.

v) checking out with session 2 (which contains 10 items) is possible.

After staging the PR, following steps i) to iii), leads to the same result as before:

![image](https://user-images.githubusercontent.com/49817236/90791037-a66eb600-e300-11ea-8fa3-d5b70eee61a0.png)

It's possible to check what's in the cart for both sessions - session 2 (right) contains the items from session 1 as well. Also, and as before, it's not possible to check out from session 1; This is only possible for session 2.

So, I could not verify the test cases and I am not sure the PR introduces a benefit. Thoughts? (adding feedback label)



Another issue found, before and after the PR, and therefore unrelated:

After step iii), if instead of attempting to checkout one attempts to:
- go back to the cart on session 1 -> a redirection to the homepage occurs
- after this, trying to checkout on session 2 is not possible, and the error "Checkout Failed. Let us know so we can process your order" is seen.

At this point, the user has items in the cart, but cannot checkout. He cannot visit /cart page - gets redirected to the homepage
Trying to visit another shop leads to the grumpy cat:

![image](https://user-images.githubusercontent.com/49817236/90792797-d74fea80-e302-11ea-8214-b5cfdd226e8f.png)

Logging out and back in does not solve the issue, so basically, this account becomes blocked, in terms of customer function. This will need an issue, as it seems like a point of no return, for the user.


Respective bugsnag error:
https://openfoodnetwork.slack.com/archives/CEF14NU3V/p1597934757001500

luisramos0 commented:
Hmmm, the interesting thing is that all this is before you stage this PR!

This PR does remove the logic of merging orders so my question is: what happens to this test case after the PR. I think it will be better! Can you please try?

luisramos0 commented:
"The two carts should be persisted and independent." this is not correct without the last commit I reverted.
When you log in to the new session your cart should be there as it was on the other session.

filipefurtad0 commented:
Hey @luisramos0,

I ran the tests before and after the PR, and got the same results... Also the issue found applies to before/after (I double checked this).

I also tried this:
- login to OFN and add something to the cart.
- Then, open an icognito tab and login again, do you see your cart as it was? -> Yup, this worked! :+1:
